[DataLoader] ConcatDataPipe

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53301 [DataLoader] ConcatDataPipe**

Differential Revision: [D26829322](https://our.internmc.facebook.com/intern/diff/D26829322)

ejguan committed:
[DataLoader] ConcatDataPipe

[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -0,0 +1,40 @@
+
+
+@functional_datapipe('concat')
+class ConcatIterDataPipe(IterDataPipe):
```
Do you guys think `ConcatIterDataPipe` is more reasonable considering BC of `ConcatDataSet` or `ChainIterDataPipe` that is similar to python itertools.

VitalyFedyunin commented on test/test_datapipe.py:
```diff
@@ -280,6 +280,27 @@ def test_picklable(self):
                 with self.assertRaises(AttributeError):
                     p = pickle.dumps(datapipe)  # type: ignore
 
+    def test_concat_datapipe(self):
```
Please add the test to iterate twice, sometimes reset doesn't work well.

ejguan committed:
Update on "[DataLoader] ConcatDataPipe"


Differential Revision: [D26829322](https://our.internmc.facebook.com/intern/diff/D26829322)

[ghstack-poisoned]
